src/github_runner_image_builder/builder.py

```diff
@@ -0,0 +1,609 @@
+            ["/usr/bin/chmod", "777", "/usr/local/bin"], check=True, timeout=30
+        )
+    except subprocess.SubprocessError as exc:
+        raise SystemUserConfigurationError from exc
+
+
+YQ_DOWNLOAD_URL_TMPL = (
+    "https://github.com/mikefarah/yq/releases/latest/download/yq_linux_{BIN_ARCH}"
+)
+YQ_BINARY_CHECKSUM_URL = "https://github.com/mikefarah/yq/releases/latest/download/checksums"
+YQ_CHECKSUM_HASHES_ORDER_URL = (
+    "https://github.com/mikefarah/yq/releases/latest/download/checksums_hashes_order"
+)
+YQ_EXTRACT_CHECKSUM_SCRIPT_URL = (
+    "https://github.com/mikefarah/yq/releases/latest/download/extract-checksum.sh"
+)
+
+
+class ExternalPackageInstallError(Exception):
+    """Represents an error installilng external packages."""
+
+
+def _validate_checksum(file: Path, expected_checksum: str) -> bool:
+    """Validate the checksum of a given file.
+
+    Args:
+        file: The file to calculate checksum for.
+        expected_checksum: The expected file checksum.
+
+    Returns:
+        True if the checksums match. False otherwise.
+    """
+    sha256 = hashlib.sha256()
+    sha256.update(file.read_bytes())
+    return sha256.hexdigest() == expected_checksum
+
+
+BIN_ARCH_MAP: dict[Arch, str] = {Arch.ARM64: "arm64", Arch.X64: "amd64"}
+
+
+def _install_external_packages(arch: Arch) -> None:
+    """Install packages outside of apt.
+
+    Installs yarn, yq.
+
+    Args:
+        arch: The architecture to download binaries for. #TODO check bin arch
+
+    Raises:
+        ExternalPackageInstallError: If there was an error installing external package.
+    """
+    try:
+        subprocess.run(
+            ["/usr/bin/npm", "install", "--global", "yarn"], check=True, timeout=60 * 5
+        )  # nosec: B603
+        subprocess.run(
+            ["/usr/bin/npm", "cache", "clean", "--force"], check=True, timeout=60
+        )  # nosec: B603
+        bin_arch = BIN_ARCH_MAP[arch]
+        yq_path_str = f"yq_linux_{bin_arch}"
+        # The URLs are trusted
+        urllib.request.urlretrieve(
+            YQ_DOWNLOAD_URL_TMPL.format(BIN_ARCH=bin_arch), yq_path_str
+        )  # nosec: B310
+        urllib.request.urlretrieve(YQ_BINARY_CHECKSUM_URL, "checksums")  # nosec: B310
+        urllib.request.urlretrieve(
+            YQ_CHECKSUM_HASHES_ORDER_URL, "checksums_hashes_order"
+        )  # nosec: B310
+        urllib.request.urlretrieve(
+            YQ_EXTRACT_CHECKSUM_SCRIPT_URL, "extract-checksum.sh"
+        )  # nosec: B310
+        # The output is <BIN_NAME> <CHECKSUM>
+        checksum = subprocess.check_output(  # nosec: B603
+            ["/usr/bin/bash", "extract-checksum.sh", "SHA-256", yq_path_str],
+            encoding="utf-8",
+            timeout=60,
+        ).split()[1]
+        yq_path = Path(yq_path_str)
+        if not _validate_checksum(yq_path, checksum):
+            raise ExternalPackageInstallError("Invalid checksum")
+        yq_path.chmod(755)
+        yq_path.rename("/usr/bin/yq")
```

Comment:
I think we might need to revisit installing yq like this in the future (due to https://github.com/canonical/is-charms-contributing-guide/blob/main/CONTRIBUTING.md#downloading-binaries )

Reply:
Good point - I remember seeing the maintainer of yq (or jq i can't recall for sure) trying to create a snap but they've been refused of making it a classic confinement and hence these applications could not be packaged.
But seems like it is actually available as a snap now: https://snapcraft.io/install/yq/ubuntu

Reply:
Since it's a snap, it might be best to have this until we can replace this in the cloud-init script :)

Reply:
Ignore my comments above - yq cannot access directories outside of $HOME still.
See issue https://forum.snapcraft.io/t/requesting-classic-confinement-for-yq/10559

Reply:
Ok, I think this is ok for now, as the scope of the application/charm is to replace the current mechanism which also downloads the binary.